src/Diagnostics.RuntimeHost/Controllers/ControllerBase.cs

```diff
@@ -93,10 +103,66 @@ protected OperationContext<TResource> PrepareContext<TResource>(TResource resour
             {
                 dataProvidersMetadata = GetDataProvidersMetadata(dataProviders);
             }
+
+            return response;
+        }
+
+        protected async Task<IEnumerable<Definition>> GetDetectorsBySupportTopicId<TResource>(OperationContext<TResource> context, string supportTopicId)
+            where TResource : IResource
+        {
+            var allDetectors = (await ListDetectors(context)).Select(detectorResponse => detectorResponse.Metadata);
+
+            var applicableDetectors = string.IsNullOrWhiteSpace(supportTopicId) ? 
+                allDetectors:
+                allDetectors.Where(detector => detector.SupportTopicList.FirstOrDefault(supportTopic => supportTopic.Id == supportTopicId) != null);
+
+            return applicableDetectors;
+        }
+
+        protected async Task<IEnumerable<AzureSupportCenterInsight>> GetInsightsFromDetector<TResource>(OperationContext<TResource> context, Definition detector)
+            where TResource: IResource
+        {
+            Response response = null;
+            try
+            {
+                response = await GetDetector(detector.Id, context);
+            }
+            catch(Exception)
+            {
+                //TODO: log
+            }
             
-            var apiResponse = DiagnosticApiResponse.FromCsxResponse(response, dataProvidersMetadata);
+            // Handle Exception or Not Found
+            // Not found can occur if invalid detector is put in detector list
+            if (response == null)
+            {
+                return null;
+            }
+
+            List<AzureSupportCenterInsight> supportCenterInsights = new List<AzureSupportCenterInsight>();
+
+            // Take max one insight per detector, only critical or warning, pick the most critical
+            var mostCriticalInsight = response.Insights.Where(insight => ((int)insight.Status) < ((int)InsightStatus.Warning)).OrderBy(insight => insight.Status).FirstOrDefault();
```

Comment:
>var mostCriticalInsight = response.Insights.Where(insight => ((int)insight.Status) < ((int)InsightStatus.Warning)).OrderBy(insight => insight.Status).FirstOrDefault(); [](start = 12, length = 167)

Why can't we show warnings and success insights? Does ASC only show problem areas?

Reply:
right now, yes. I actually was going to make it a query string parameter to filter only critical/warning. Still need to get from them what the enum values they have for status

Reply:
Yea query string parameter sounds like a good idea to save yourself a deployment

---
In reply to: [193861176](https://github.com/Azure/Azure-AppServices-Diagnostics/pull/39#discussion_r193861176) [](ancestors = 193861176)